community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -626,7 +623,17 @@ private void afterRollback()
     private void release()
     {
         locks.releaseAll();
-        pool.release( this );
+        if ( terminated )
+        {
+            // This transaction has been externally marked for termination.
+            // Just dispose of this transaction and don't return it to the pool.
+            dispose();
+        }
+        else
+        {
+            // Return this instance to the pool so that another transaction may use it.
+            pool.release( this );
+        }
```

Comment:
I agree with what you say and we _could_ do that, although it's additional complexity which there's quite little value in. Releasing terminated transactions to be GCed, even those those were "just" cancelled should be OK, shouldn't it?